spng.c

```diff
@@ -2538,26 +2583,53 @@ int spng_decoded_image_size(spng_ctx *ctx, int fmt, size_t *out)
     int ret = get_ancillary(ctx);
     if(ret) return ret;
 
-    size_t res;
-    if(fmt == SPNG_FMT_RGBA8)
+    /* define parameters used to calculate total uncompressed bytes */
+    size_t pixel_size;
+    if(fmt == SPNG_FMT_RAW)
     {
-        if(4 > SIZE_MAX / ctx->ihdr.width) return SPNG_EOVERFLOW;
-        res = 4 * ctx->ihdr.width;
-
-        if(res > SIZE_MAX / ctx->ihdr.height) return SPNG_EOVERFLOW;
-        res = res * ctx->ihdr.height;
+        pixel_size = ctx->ihdr.bit_depth;
+        switch ((enum spng_color_type)ctx->ihdr.color_type) {
+            case SPNG_COLOR_TYPE_GRAYSCALE:
+            case SPNG_COLOR_TYPE_INDEXED:
+                break;
+            case SPNG_COLOR_TYPE_TRUECOLOR:
+                pixel_size *= 3;
+                break;
+            case SPNG_COLOR_TYPE_GRAYSCALE_ALPHA:
+                pixel_size *= 2;
+                break;
+            case SPNG_COLOR_TYPE_TRUECOLOR_ALPHA:
+                pixel_size *= 4;
+                break;
+        }
+    }
+    else if(fmt == SPNG_FMT_RGBA8)
+    {
+        pixel_size = 32;
     }
     else if(fmt == SPNG_FMT_RGBA16)
     {
-        if(8 > SIZE_MAX / ctx->ihdr.width) return SPNG_EOVERFLOW;
-        res = 8 * ctx->ihdr.width;
-
-        if(res > SIZE_MAX / ctx->ihdr.height) return SPNG_EOVERFLOW;
-        res = res * ctx->ihdr.height;
+        pixel_size = 64;
     }
     else return SPNG_EFMT;
 
-    *out = res;
+    /* For grayscale data with bit depths less than 8 the total number of
+    bytes cannot exceed SIZE_MAX==2**64-1, even with an alpha channel */
+    if (pixel_size < 8) {
+        if (SIZE_MAX / ctx->ihdr.width < ctx->ihdr.height) {
+            return SPNG_EOVERFLOW;
+        }
+        *out = (ctx->ihdr.width * pixel_size + 7) >> 3;
```

Comment:
This can overflow on 32-bit platforms, e.g. `png_u32max * 4` or `2147483647  * 4`.